pytrickle/frames.py

```diff
@@ -179,4 +244,235 @@ def with_monotonic_timestamps(cls, frames: List[AudioFrame], request_id: str, st
+            "total_received": self.total_frames_received,
+            "total_discarded": self.total_frames_discarded
+        }
+
+
+class StreamState:
+    """Unified state management for stream lifecycle."""
+    
+    def __init__(self):
+        self.running = False
+        self.pipeline_ready = False
+        self.shutting_down = False
+        self.error_occurred = False
+        self.cleanup_in_progress = False
+        
+        self.running_event = asyncio.Event()
+        self.shutdown_event = asyncio.Event()
+        self.error_event = asyncio.Event()
+        self.pipeline_ready_event = asyncio.Event()
+    
+    @property
+    def is_active(self) -> bool:
+        return self.running and not self.shutting_down and not self.error_occurred
+    
+    @property
+    def shutdown_flags(self) -> Dict[str, bool]:
+        return {
+            'shutdown_event': self.shutdown_event.is_set(),
+            'cleanup_in_progress': self.cleanup_in_progress
+        }
+    
+    def start(self):
+        self.running = True
+        self.running_event.set()
+    
+    def mark_pipeline_ready(self):
+        self.pipeline_ready = True
+        self.pipeline_ready_event.set()
+    
+    def initiate_shutdown(self, due_to_error: bool = False):
+        self.shutting_down = True
+        self.shutdown_event.set()
+        if due_to_error:
+            self.error_occurred = True
+            self.error_event.set()
+    
+    def mark_cleanup_in_progress(self):
+        self.cleanup_in_progress = True
+    
+    def finalize(self):
+        self.running = False
+        self.running_event.clear()
+
+
+class StreamErrorHandler:
+    """Centralized error handling for streaming applications."""
+    
+    @staticmethod
+    def log_error(error_type: str, exception: Optional[Exception], request_id: str, critical: bool = False):
+        level = logger.error if critical else logger.warning
+        msg = f"{error_type} for stream {request_id}"
+        if exception:
+            msg += f": {exception}"
+        level(msg)
+
+    @staticmethod
+    def is_shutdown_error(shutdown_flags: Dict) -> bool:
+        return shutdown_flags.get('shutdown_event', False) or shutdown_flags.get('cleanup_in_progress', False)
+
+
+class StreamingUtils:
+    """Generic utilities for streaming applications."""
+    
+    @staticmethod
+    async def cancel_task_with_timeout(task: Optional[asyncio.Task], task_name: str, timeout: float = 3.0) -> bool:
+        """Cancel an asyncio task with a timeout."""
+        if not task or task.done():
+            return True
+        task.cancel()
+        try:
+            await asyncio.wait_for(task, timeout=timeout)
+            return True
+        except (asyncio.CancelledError, asyncio.TimeoutError):
+            return True
+        except Exception:
+            return False
+
+
+# Legacy alias for backward compatibility
+def tensor_to_av_frame_legacy(tensor: torch.Tensor) -> av.VideoFrame:
```

Comment:
Is this needed?

Reply:
Removed as it is unused and no backward compatibility methods needed